src/Core/Resolvers/SqlPaginationUtil.cs

```diff
@@ -29,16 +30,37 @@ public static class SqlPaginationUtil
         /// <list>*Connection.hasNextPage which is decided on whether structure.Limit() elements have been returned</list>
         /// </list>
         /// </summary>
-        public static JsonDocument CreatePaginationConnectionFromJsonElement(JsonElement root, PaginationMetadata paginationMetadata)
+        public static JsonElement CreatePaginationConnectionFromJsonElement(JsonElement root, PaginationMetadata paginationMetadata)
+            => CreatePaginationConnection(root, paginationMetadata).ToJsonElement();
+
+        /// <summary>
+        /// Wrapper for CreatePaginationConnectionFromJsonElement
+        /// Disposes the JsonDocument passed to it
+        /// </summary>
+        public static JsonDocument CreatePaginationConnectionFromJsonDocument(JsonDocument? jsonDocument, PaginationMetadata paginationMetadata)
+        {
+            // necessary for MsSql because it doesn't coalesce list query results like Postgres
+            if (jsonDocument is null)
+            {
+                jsonDocument = JsonDocument.Parse("[]");
+            }
+
+            JsonElement root = jsonDocument.RootElement.Clone();
+
+            // create the connection object.
+            return CreatePaginationConnection(root, paginationMetadata).ToJsonDocument();
+        }
+
+        private static JsonObject CreatePaginationConnection(JsonElement root, PaginationMetadata paginationMetadata)
         {
             // Maintains the connection JSON object *Connection
-            Dictionary<string, object> connectionJson = new();
+            JsonObject connection = new();
 
             // in dw we wrap array with "" and hence jsonValueKind is string instead of array.
             if (root.ValueKind is JsonValueKind.String)
             {
                 JsonDocument document = JsonDocument.Parse(root.GetString()!);
-                root = document.RootElement;
+                root = document.RootElement.Clone();
```

Comment:
does this document need to be disposed off?

Reply:
added a using because this jsondocument is essentially self contained here and the relevant jsonelement cloned